src/contracts/atlas/Atlas.sol

```diff
@@ -246,47 +247,67 @@ contract Atlas is Escrow, Factory {
+        // position. This reduces operations needed later when sorting the array in ascending order.
+        // Each non-zero bid amount is packed with its original solverOps array index, to fit into a uint256 value. The
+        // order of bidAmount and index is important - with bidAmount using the most significant bits, and considering
+        // we do not store zero bids in the array, the index values within the uint256 should not impact the sorting.
+
+        // |<------------------------- uint256 (256 bits) ------------------------->|
+        // |                                                                        |
+        // |<------------------ uint240 ----------------->|<-------- uint16 ------->|
+        // |                                              |                         |
+        // |                    bidAmount                 |          index          |
+        // |                                              |                         |
+        // |<------------------ 240 bits ---------------->|<------- 16 bits ------->|
+
+        for (uint256 i; i < solverOps.length; ++i) {
+            bidAmountFound = _getBidAmount(dConfig, userOp, solverOps[i], returnData, key);
+
+            if (bidAmountFound == 0 || bidAmountFound > type(uint240).max) {
+                // Zero bids are ignored: increment zeroBidCount offset
+                // Bids that would cause an overflow are also ignored
                 unchecked {
-                    ++j;
+                    ++zeroBidCount;
                 }
-                continue;
             } else {
-                bidAmounts[i] = bidPlaceholder;
-
-                for (uint256 k = i - j + 1; k > 0; k--) {
-                    if (bidPlaceholder > bidAmounts[sortedOps[k - 1]]) {
-                        sortedOps[k] = sortedOps[k - 1];
-                        sortedOps[k - 1] = i;
-                    } else {
-                        sortedOps[k] = i;
-                        break;
-                    }
-                }
+                // Non-zero bids are packed with their original solverOps index.
+                // The array is filled with non-zero bids from the right. This causes all zero bids to be on the left -
+                // in their sorted position, so fewer operations are needed in the sorting step below.
+                bidsAndIndices[bidsAndIndicesLastIndex - (i - zeroBidCount)] =
+                    uint256(bidAmountFound << BITS_FOR_INDEX | uint16(i));
             }
         }
 
+        // Then, sorts the uint256 array in-place, in ascending order.
+        LibSort.insertionSort(bidsAndIndices);
+
         key.bidFind = false;
-        j = solverOps.length - j;
 
-        for (uint256 i; i < j; i++) {
-            bidPlaceholder = sortedOps[i];
+        // Finally, iterate through sorted bidsAndIndices array in descending order of bidAmount.
+        for (uint256 i = bidsAndIndices.length - 1; i >= 0; --i) {
```

Comment:
I think here is an issue: when `i==0` then `i >= 0` is still true and it tries to decrease `i`, which will fail.

Reply:
An Ex Post Bids Atlas transaction should never be created by the bundler if there are 0 bids I believe.

Reply:
@gpersoon good catch, will fix that
@jgcogsystematictrading I think he means bid amounts of zero, found on-chain using the whole bid finding system with ex post bids. Not 0 solverOps.